Class not found when using java.time classes in @ElementCollection

**Describe the bug**
I built a simple application with Quarkus 0.21.2 which uses hibernate orm panach to save a entity to a h2 database. The entity includes a ElementCollection of DaysOfWeek. I also created some tests to ensure CURD works. The tests all work fine but when I run them nativly I get the following exception:
```
Exception in thread "main" java.lang.RuntimeException: Failed to start quarkus
    at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:126)
    at io.quarkus.runtime.Application.start(Application.java:91)
    at io.quarkus.runtime.Application.run(Application.java:204)
    at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:34)
Caused by: javax.persistence.PersistenceException: [PersistenceUnit: default] Unable to build Hibernate SessionFactory
    at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.persistenceException(FastBootEntityManagerFactoryBuilder.java:113)
    at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:67)
    at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:54)
    at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:79)
    at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:54)
    at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:109)
    at io.quarkus.hibernate.orm.runtime.JPAConfig.startAll(JPAConfig.java:57)
    at io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder.startAllPersistenceUnits(HibernateOrmRecorder.java:77)
    at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits19.deploy_0(HibernateOrmProcessor$startPersistenceUnits19.zig:51)
    at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits19.deploy(HibernateOrmProcessor$startPersistenceUnits19.zig:70)
    at io.quarkus.runner.ApplicationImpl1.doStart(ApplicationImpl1.zig:90)
    ... 3 more
Caused by: org.hibernate.MappingException: Could not create DynamicParameterizedType for type: org.hibernate.type.EnumType
    at org.hibernate.mapping.SimpleValue.createParameterImpl(SimpleValue.java:768)
    at org.hibernate.mapping.SimpleValue.getType(SimpleValue.java:470)
    at org.hibernate.mapping.SimpleValue.isValid(SimpleValue.java:455)
    at org.hibernate.mapping.Collection.validate(Collection.java:315)
    at org.hibernate.boot.internal.MetadataImpl.validate(MetadataImpl.java:347)
    at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:461)
    at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:65)
    ... 12 more
Caused by: org.hibernate.boot.registry.classloading.spi.ClassLoadingException: Unable to load class [java.time.DayOfWeek]
    at io.quarkus.hibernate.orm.runtime.service.FlatClassLoaderService.classForName(FlatClassLoaderService.java:39)
    at org.hibernate.mapping.SimpleValue.createParameterImpl(SimpleValue.java:755)
    ... 18 more
Caused by: java.lang.ClassNotFoundException: java.time.DayOfWeek
    at com.oracle.svm.core.hub.ClassForNameSupport.forName(ClassForNameSupport.java:60)
    at java.lang.Class.forName(DynamicHub.java:1174)
    at io.quarkus.hibernate.orm.runtime.service.FlatClassLoaderService.classForName(FlatClassLoaderService.java:37)
    ... 19 more
```

I wrote about this on StackOverflow before and got a Workaround and got quested to post it as a issue here too: https://stackoverflow.com/questions/57946841/quarkus-test-fails-in-native-becaus-dayofweek-not-found

What I found strange as I tried the workaround was, that the text file with the used classes (`used_classes_events-service-1.0-SNAPSHOT-runner_20190917_084733.txt`) which was generated while creating the native runner already included the `java.time.DayOfWeek`. But the exception was only fixed when I added the workaround.

**Expected behavior**
Tests run successfully like the non native tests.

**Actual behavior**
The native tests errored with an exception.


**To Reproduce**
I published my whole project on github so you have all you need to reproduce:
https://github.com/Nicklas2751/quarkus-issue-example

Steps to reproduce the behavior:
1. `./mvnw verify -P native`

**Configuration**
src/main/resources/application.properties
```properties
# configure your datasource
quarkus.datasource.url=jdbc:h2:tcp://localhost/mem:default
quarkus.datasource.driver=org.h2.Driver
quarkus.datasource.username=username-default
quarkus.datasource.min-size=3
quarkus.datasource.max-size=13
# drop and create the database at startup (use `update` to only update the schema)
quarkus.hibernate-orm.database.generation=drop-and-create
```

src/test/resources/application.properties
```properties
quarkus.datasource.url=jdbc:h2:tcp://localhost/mem:test
quarkus.datasource.driver=org.h2.Driver
quarkus.hibernate-orm.database.generation=drop-and-create
```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Linux CL003072 4.15.0-62-generic #69-Ubuntu SMP Wed Sep 4 20:55:53 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

 - Output of `java -version`: openjdk version "1.8.0_212"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_212-b03)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.212-b03, mixed mode)

 - GraalVM version (if different from Java): 19.2.0
 - Quarkus version or git rev: 0.21.2
```

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java
@@ -1,6 +1,7 @@
 package io.quarkus.hibernate.orm.deployment;
 
 import java.io.IOException;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.List;
@@ -8,6 +9,7 @@ import java.util.Set;
 import java.util.TreeSet;
 import java.util.stream.Collectors;
 
+import javax.persistence.ElementCollection;
 import javax.persistence.Embeddable;
 import javax.persistence.Embedded;
 import javax.persistence.Entity;
@@ -21,6 +23,7 @@ import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.FieldInfo;
 import org.jboss.jandex.IndexView;
+import org.jboss.jandex.Type;
 
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
@@ -41,7 +44,9 @@ final class JpaJandexScavenger {
 
     private static final DotName JPA_ENTITY = DotName.createSimple(Entity.class.getName());
     private static final DotName EMBEDDABLE = DotName.createSimple(Embeddable.class.getName());
-    private static final DotName EMBEDDED = DotName.createSimple(Embedded.class.getName());
+    private static final List<DotName> EMBEDDED_ANNOTATIONS = Arrays.asList(
+            DotName.createSimple(Embedded.class.getName()),
+            DotName.createSimple(ElementCollection.class.getName()));
     private static final DotName MAPPED_SUPERCLASS = DotName.createSimple(MappedSuperclass.class.getName());
 
     private static final DotName ENUM = DotName.createSimple(Enum.class.getName());
@@ -69,15 +74,21 @@ final class JpaJandexScavenger {
         // Not functional as we will need one deployment template per persistence unit
         final JpaEntitiesBuildItem domainObjectCollector = new JpaEntitiesBuildItem();
         final Set<String> enumTypeCollector = new HashSet<>();
+        final Set<String> javaTypeCollector = new HashSet<>();
         final Set<DotName> unindexedClasses = new TreeSet<>();
 
-        enlistJPAModelClasses(indexView, domainObjectCollector, enumTypeCollector, JPA_ENTITY, unindexedClasses);
-        enlistJPAModelClasses(indexView, domainObjectCollector, enumTypeCollector, EMBEDDABLE, unindexedClasses);
-        enlistJPAModelClasses(indexView, domainObjectCollector, enumTypeCollector, MAPPED_SUPERCLASS, unindexedClasses);
-        enlistReturnType(indexView, domainObjectCollector, enumTypeCollector, unindexedClasses);
+        enlistJPAModelClasses(indexView, domainObjectCollector, enumTypeCollector, javaTypeCollector, JPA_ENTITY,
+                unindexedClasses);
+        enlistJPAModelClasses(indexView, domainObjectCollector, enumTypeCollector, javaTypeCollector, EMBEDDABLE,
+                unindexedClasses);
+        enlistJPAModelClasses(indexView, domainObjectCollector, enumTypeCollector, javaTypeCollector, MAPPED_SUPERCLASS,
+                unindexedClasses);
+        enlistEmbeddedsAndElementCollections(indexView, domainObjectCollector, enumTypeCollector, javaTypeCollector,
+                unindexedClasses);
 
         for (PersistenceUnitDescriptor pud : explicitDescriptors) {
-            enlistExplicitClasses(indexView, domainObjectCollector, enumTypeCollector, pud.getManagedClassNames(),
+            enlistExplicitClasses(indexView, domainObjectCollector, enumTypeCollector, javaTypeCollector,
+                    pud.getManagedClassNames(),
                     unindexedClasses);
         }
 
@@ -88,6 +99,12 @@ final class JpaJandexScavenger {
             for (String className : enumTypeCollector) {
                 reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, className));
             }
+        }
+
+        // for the java. types we collected (usually from java.time but it could be from other types),
+        // we just register them for reflection
+        for (String javaType : javaTypeCollector) {
+            reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, javaType));
         }
 
         if (!unindexedClasses.isEmpty()) {
@@ -110,7 +127,8 @@ final class JpaJandexScavenger {
     }
 
     private static void enlistExplicitClasses(IndexView index, JpaEntitiesBuildItem domainObjectCollector,
-            Set<String> enumTypeCollector, List<String> managedClassNames, Set<DotName> unindexedClasses) {
+            Set<String> enumTypeCollector, Set<String> javaTypeCollector, List<String> managedClassNames,
+            Set<DotName> unindexedClasses) {
         for (String className : managedClassNames) {
             DotName dotName = DotName.createSimple(className);
             boolean isInIndex = index.getClassByName(dotName) != null;
@@ -118,39 +136,44 @@ final class JpaJandexScavenger {
                 unindexedClasses.add(dotName);
             }
 
-            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, dotName, unindexedClasses);
+            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, javaTypeCollector, dotName,
+                    unindexedClasses);
         }
     }
 
-    private static void enlistReturnType(IndexView index, JpaEntitiesBuildItem domainObjectCollector,
-            Set<String> enumTypeCollector, Set<DotName> unindexedClasses) {
-        Collection<AnnotationInstance> annotations = index.getAnnotations(EMBEDDED);
-        if (annotations == null) {
-            return;
+    private static void enlistEmbeddedsAndElementCollections(IndexView index, JpaEntitiesBuildItem domainObjectCollector,
+            Set<String> enumTypeCollector, Set<String> javaTypeCollector, Set<DotName> unindexedClasses) {
+        Set<DotName> embeddedTypes = new HashSet<>();
+
+        for (DotName embeddedAnnotation : EMBEDDED_ANNOTATIONS) {
+            Collection<AnnotationInstance> annotations = index.getAnnotations(embeddedAnnotation);
+
+            for (AnnotationInstance annotation : annotations) {
+                AnnotationTarget target = annotation.target();
+
+                switch (target.kind()) {
+                    case FIELD:
+                        collectEmbeddedTypes(embeddedTypes, target.asField().type());
+                        break;
+                    case METHOD:
+                        collectEmbeddedTypes(embeddedTypes, target.asMethod().returnType());
+                        break;
+                    default:
+                        throw new IllegalStateException(
+                                "[internal error] " + embeddedAnnotation + " placed on a unknown element: " + target);
+                }
+
+            }
         }
 
-        for (AnnotationInstance annotation : annotations) {
-            AnnotationTarget target = annotation.target();
-            DotName jpaClassName;
-            switch (target.kind()) {
-                case FIELD:
-                    // TODO could fail if that's an array or a generic type
-                    jpaClassName = target.asField().type().name();
-                    break;
-                case METHOD:
-                    // TODO could fail if that's an array or a generic type
-                    jpaClassName = target.asMethod().returnType().name();
-                    break;
-                default:
-                    throw new IllegalStateException("[internal error] @Embedded placed on a unknown element: " + target);
-            }
-            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, jpaClassName,
+        for (DotName embeddedType : embeddedTypes) {
+            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, javaTypeCollector, embeddedType,
                     unindexedClasses);
         }
     }
 
     private void enlistJPAModelClasses(IndexView index, JpaEntitiesBuildItem domainObjectCollector,
-            Set<String> enumTypeCollector, DotName dotName, Set<DotName> unindexedClasses) {
+            Set<String> enumTypeCollector, Set<String> javaTypeCollector, DotName dotName, Set<DotName> unindexedClasses) {
         Collection<AnnotationInstance> jpaAnnotations = index.getAnnotations(dotName);
 
         if (jpaAnnotations == null) {
@@ -164,7 +187,7 @@ final class JpaJandexScavenger {
             if (nonJpaModelClasses.contains(targetDotName.toString())) {
                 continue;
             }
-            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, targetDotName,
+            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, javaTypeCollector, targetDotName,
                     unindexedClasses);
             collectDomainObject(domainObjectCollector, klass);
         }
@@ -178,9 +201,15 @@ final class JpaJandexScavenger {
      * TODO should we also return the return types of all methods and fields? It could contain Enums for example.
      */
     private static void addClassHierarchyToReflectiveList(IndexView index, JpaEntitiesBuildItem domainObjectCollector,
-            Set<String> enumTypeCollector, DotName className, Set<DotName> unindexedClasses) {
-        if (className == null || className.toString().startsWith("java.")) {
-            // bail out if java.lang.Object or any java. class
+            Set<String> enumTypeCollector, Set<String> javaTypeCollector, DotName className, Set<DotName> unindexedClasses) {
+        if (className == null || isIgnored(className)) {
+            // bail out if java.lang.Object or a class we want to ignore
+            return;
+        }
+
+        // if the class is in the java. package and is not ignored, we want to register it for reflection
+        if (isInJavaPackage(className)) {
+            javaTypeCollector.add(className.toString());
             return;
         }
 
@@ -202,11 +231,13 @@ final class JpaJandexScavenger {
         collectDomainObject(domainObjectCollector, classInfo);
 
         // add superclass recursively
-        addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, classInfo.superName(),
+        addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, javaTypeCollector,
+                classInfo.superName(),
                 unindexedClasses);
         // add interfaces recursively
         for (DotName interfaceDotName : classInfo.interfaceNames()) {
-            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, interfaceDotName,
+            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, javaTypeCollector,
+                    interfaceDotName,
                     unindexedClasses);
         }
     }
@@ -217,5 +248,41 @@ final class JpaJandexScavenger {
         } else {
             domainObjectCollector.addModelClass(modelClass.name().toString());
         }
+    }
+
+    private static void collectEmbeddedTypes(Set<DotName> embeddedTypes, Type indexType) {
+        switch (indexType.kind()) {
+            case CLASS:
+                embeddedTypes.add(indexType.asClassType().name());
+                break;
+            case PARAMETERIZED_TYPE:
+                embeddedTypes.add(indexType.name());
+                for (Type typeArgument : indexType.asParameterizedType().arguments()) {
+                    collectEmbeddedTypes(embeddedTypes, typeArgument);
+                }
+                break;
+            case ARRAY:
+                collectEmbeddedTypes(embeddedTypes, indexType.asArrayType().component());
+                break;
+            default:
+                // do nothing
+                break;
+        }
+    }
+
+    private static boolean isIgnored(DotName classDotName) {
+        String className = classDotName.toString();
+        if (className.startsWith("java.util.") || className.startsWith("java.lang.")) {
+            return true;
+        }
+        return false;
+    }
+
+    private static boolean isInJavaPackage(DotName classDotName) {
+        String className = classDotName.toString();
+        if (className.startsWith("java.")) {
+            return true;
+        }
+        return false;
     }
 }

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/elementcollection/ElementCollectionResource.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/elementcollection/ElementCollectionResource.java
@@ -1,0 +1,32 @@
+package io.quarkus.it.jpa.elementcollection;
+
+import java.time.DayOfWeek;
+import java.time.LocalTime;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import javax.transaction.Transactional;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/element-collection")
+@ApplicationScoped
+public class ElementCollectionResource {
+
+    @Inject
+    EntityManager em;
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    @Transactional
+    public String create() {
+        OpeningTimes openingTimes = new OpeningTimes("standard", LocalTime.of(10, 0), LocalTime.of(19, 0), DayOfWeek.MONDAY,
+                DayOfWeek.TUESDAY, DayOfWeek.THURSDAY, DayOfWeek.FRIDAY);
+        em.persist(openingTimes);
+
+        return "OK";
+    }
+}

--- a/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/elementcollection/OpeningTimes.java
+++ b/integration-tests/jpa/src/main/java/io/quarkus/it/jpa/elementcollection/OpeningTimes.java
@@ -1,0 +1,128 @@
+package io.quarkus.it.jpa.elementcollection;
+
+import static org.hibernate.annotations.CascadeType.ALL;
+
+import java.time.DayOfWeek;
+import java.time.LocalTime;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+
+import javax.persistence.ElementCollection;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.validation.constraints.NotNull;
+
+import org.hibernate.annotations.Cascade;
+import org.hibernate.annotations.GenericGenerator;
+
+@Entity
+public class OpeningTimes {
+    @Id
+    @GeneratedValue(generator = "UUID")
+    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
+    private UUID id;
+
+    @NotNull
+    private String name;
+    private String description;
+    private LocalTime timeFrom;
+    private LocalTime timeTo;
+
+    @ElementCollection
+    @Cascade(value = { ALL })
+    private Collection<DayOfWeek> daysOfWeek;
+
+    OpeningTimes() {
+        name = "";
+    }
+
+    public OpeningTimes(
+            String name,
+            LocalTime timeFrom,
+            LocalTime timeTo,
+            Collection<DayOfWeek> daysOfWeek) {
+        this.name = name;
+        this.timeFrom = timeFrom;
+        this.timeTo = timeTo;
+        this.daysOfWeek = new HashSet<>(daysOfWeek);
+    }
+
+    public OpeningTimes(String name, LocalTime from, LocalTime to, DayOfWeek... daysOfWeek) {
+        this(name, from, to, new ArrayList<>(Arrays.asList(daysOfWeek)));
+    }
+
+    public LocalTime getTimeFrom() {
+        return timeFrom;
+    }
+
+    public void setTimeFrom(LocalTime timeFrom) {
+        this.timeFrom = timeFrom;
+    }
+
+    public LocalTime getTimeTo() {
+        return timeTo;
+    }
+
+    public void setTimeTo(LocalTime timeTo) {
+        this.timeTo = timeTo;
+    }
+
+    public Collection<DayOfWeek> getDaysOfWeek() {
+        return daysOfWeek;
+    }
+
+    public void setDaysOfWeek(Set<DayOfWeek> daysOfWeek) {
+        this.daysOfWeek = daysOfWeek;
+    }
+
+    public UUID getId() {
+        return id;
+    }
+
+    public void setId(UUID id) {
+        this.id = id;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    public void setDescription(String description) {
+        this.description = description;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (!(o instanceof OpeningTimes))
+            return false;
+        OpeningTimes that = (OpeningTimes) o;
+        return Objects.equals(getId(), that.getId())
+                && Objects.equals(getName(), that.getName())
+                && Objects.equals(getDescription(), that.getDescription())
+                && Objects.equals(getTimeFrom(), that.getTimeFrom())
+                && Objects.equals(getTimeTo(), that.getTimeTo())
+                && Objects.equals(getDaysOfWeek(), that.getDaysOfWeek());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(
+                getId(), getName(), getDescription(), getTimeFrom(), getTimeTo(), getDaysOfWeek());
+    }
+}

--- a/integration-tests/jpa/src/test/java/io/quarkus/it/jpa/configurationless/JPAElementCollectionInGraalITCase.java
+++ b/integration-tests/jpa/src/test/java/io/quarkus/it/jpa/configurationless/JPAElementCollectionInGraalITCase.java
@@ -1,0 +1,8 @@
+package io.quarkus.it.jpa.configurationless;
+
+import io.quarkus.test.junit.SubstrateTest;
+
+@SubstrateTest
+public class JPAElementCollectionInGraalITCase extends JPAElementCollectionTest {
+
+}

--- a/integration-tests/jpa/src/test/java/io/quarkus/it/jpa/configurationless/JPAElementCollectionTest.java
+++ b/integration-tests/jpa/src/test/java/io/quarkus/it/jpa/configurationless/JPAElementCollectionTest.java
@@ -1,0 +1,19 @@
+package io.quarkus.it.jpa.configurationless;
+
+import static org.hamcrest.core.StringContains.containsString;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.RestAssured;
+
+@QuarkusTest
+public class JPAElementCollectionTest {
+
+    @Test
+    public void testInjection() {
+        RestAssured.when().get("/jpa-test/element-collection").then()
+                .statusCode(200)
+                .body(containsString("OK"));
+    }
+}

